cachi2/core/package_managers/yarn/project.py

```diff
@@ -7,58 +7,43 @@
 import json
 import logging
 import re
+from collections import UserDict
 from pathlib import Path
 from typing import Any, NamedTuple, Optional
 
 import semver
 import yaml
 
 from cachi2.core.errors import PackageRejected, UnexpectedFormat
+from cachi2.core.package_managers.yarn.utils import run_yarn_cmd
 from cachi2.core.rooted_path import RootedPath
 
 log = logging.getLogger(__name__)
 
 
-DEFAULT_CACHE_FOLDER = "./.yarn/cache"
-DEFAULT_REGISTRY = "https://registry.yarnpkg.com"
-
-
-class YarnRc:
+class YarnRc(UserDict):
     """A yarnrc file.
 
-    This class abstracts the underlying attributes and only exposes what
-    is relevant for the request processing.
+    This class abstracts the underlying attributes of a .yarnrc YAML
+    configuration file.
     """
 
-    def __init__(self, path: RootedPath, data: dict[str, Any]) -> None:
-        """Initialize a YarnRc object.
+    def __init__(
+        self, data: dict[str, Any] = {}, defaults: Optional[dict[str, Any]] = None
+    ) -> None:
+        """Initialize a YarnRc dictionary.
 
-        :param path: the path to the yarnrc file, relative to the request source dir.
         :param data: the raw data for the yarnrc file.
         """
-        self._path = path
```

Comment:
> why should the data be written to a static location on the disk rather than making the interface more flexible and let the caller specify where the config should be dumped to.

Because then the caller doesn't need to have knowledge of the file location in the disk :stuck_out_tongue:  

We'll be calling `write` on yarn/main.py, and I don't think it should know/care about the file locations. We could give this responsibility to the `Project` class, though, if you want to make `write` more flexible.

Reply:
> Because then the caller doesn't need to have knowledge of the file location in the disk 

That's the thing though. A config map is just an in-memory representation with an optional bunch of utility methods, but since the only use case for the "knowledge" of the destination path of the config is for the `write` utility method it feels like this particular class isn't the right place to store it. Moreover, you need to know upfront where you'd want to dump the RC file to when you instantiate the class. What if you need to dump it elsewhere, i.e. not to the usual location (e.g. during unit tests)? From that POV to me it makes sense to push the ownership of the disk location where it makes more sense - the caller.

> We could give this responsibility to the Project class, though, if you want to make write more flexible.

Agreed, this is what the term `caller` in the commit message was originally meant to refer to. :)